### PR TITLE
sub: refactor to use gangway

### PR DESCRIPTION
**Pull request by listx**
This de-duplicates logic in Sub to use Gangway's idea of a `CreateJobExecutionRequest`. It does this by converting the incoming PubSub payload into a CreateJobExecutionRequest in msgToCjer().

/cc @chaodaiG @mpherman2

### CODE REVIEW

**Comment by listx** [comment]
FTR the changes in this PR were extracted from the larger (closed) PR here: https://github.com/kubernetes/test-infra/pull/28036

The only difference is that I fixed a typo in one of the comments, but otherwise it's the same.

**Review by chaodaiG — COMMENTED** [review]
generally LGTM, a couple of comments

File: prow/pubsub/subscriber/subscriber.go [hunk 1]
```diff
@@ -422,116 +187,91 @@ func (s *Subscriber) handleMessage(msg messageInterface, subscription string, al
 	return err
 }
 
-func (s *Subscriber) handleProwJob(l *logrus.Entry, jh jobHandler, msgPayload []byte, subscription, eType string, allowedClusters []string) error {
+// msgToCjer converts an incoming message (PubSub message) into a CJER. It
+// actually does 2 conversions --- from the message to ProwJobEvent (in order to
+// unmarshal the raw bytes) then again from ProwJobEvent to a CJER.
+func (s *Subscriber) msgToCjer(l *logrus.Entry, msg messageInterface, subscription string) (*gangway.CreateJobExecutionRequest, error) {
```
[the remark this answers is not in the document]
Comment (listx) [reply prow/pubsub/subscriber/subscriber.go:193]: But isn't Sub the only component where it deals with PubSub messages? Moving it outside would make more sense if other things also need to use `msgToCjer`. Even if we did move it out to a different file, that file would still probably have to live under `prow/pubsub/...` right?

File: prow/pubsub/subscriber/subscriber.go [hunk 2]
```diff
@@ -422,116 +187,91 @@ func (s *Subscriber) handleMessage(msg messageInterface, subscription string, al
...
-	combinedAnnotations := make(map[string]string)
-	for k, v := range labels {
-		combinedLabels[k] = v
-	}
-	for k, v := range annotations {
-		combinedAnnotations[k] = v
+	// First encode the job type.
+	switch eType {
+	case PeriodicProwJobEvent:
+		cjer.JobExecutionType = gangway.JobExecutionType_PERIODIC
```
[the remark this answers is not in the document]
Comment (listx) [reply prow/pubsub/subscriber/subscriber.go:235]: I think so, but then I'd have to make the refactor more aggressive. IIRC the reason I did this was to do the minimal amount of work to make Sub use Gangway code. If we skip the use of a `ProwJobEvent` then the methods that use it as a pointer receive near the top of this file all have to be rewritten, probably.

I think that could be done as a lower-priority follow-up PR.

**Comment by listx** [comment]
@chaodaiG are your comments blockers? If not, can we get this merged so that @timwangmusic is not blocked (one of their starter projects is for improving Sub).

**Review by chaodaiG — APPROVED** [review]

**Pull request merged**